### ISSUE TITLE
Formally Support Games Distributing Their Own Clients

### ISSUE THREAD

**Issue by benrob0329** [minetest/minetest#8727]
## Problem

Note: This is a **Discussion** Issue, it is for discussing whether this is wanted or feasible.

Minetest is a game engine, but one which acts like a game. This can be problematic for game developers who want their game to be respected as independent, or who want further control over the client configuration. I feel that this issue has been touched upon over the years, but no one has ever officially stated whether or not that's needed or wanted for MT.

For example, `subgames` were renamed to `games` in order to further the idea that they are independent projects and more than just large mods, but they still live within the Minetest ecosystem and are still fancy mods for this strange Gmod-style game we call an engine. So while we can add a catch-all content store, and while we can add menu customization and background music, it doesn't get away from the fact that games made with MT will always require users to install MT first, then the game through the content store.

It will always _feel_ like a subgame, even if we've renamed it.

## Solutions

We could make Minetest into something akin to Love2D, a Lua runtime environment with all the tools needed to make and distribute a game. This would be the easiest and most logical approach to it, as it's what most Lua game engines do. The CDB and serverlist would most likely become game-specific software that the developer's would host for their games and related content.

We could also radically change the structure of Minetest and make it into a full library that runs on the system with hooks that people can use, but that sounds like a major refactoring of the entire project that will never happen and probably shouldn't to begin with.

## Alternatives

We could do nothing at all. We could keep this strange mixture of Game and Engine, after all most games and mods for MT are centered around MTG, and those that aren't are usually small and don't have the resources to package and distribute their game anyways.

But are we encouraging that by keeping it this way?

**EDIT: We could also push towards being a content distribution platform, however this isn't quite what most would consider a game engine.**

## Additional context

I'm somewhat neutral on this, but I feel that it needs to be discussed because people are both distributing their own clients, as well as packaging thing in the CDB.

**Comment by Linuxdirk** [comment]
> Minetest is a game engine, but one which acts like a game.

No, it doesn’t. It is a game engine. But the engine gets shipped with a boring and uninspired non-game as default.

> The CDB and serverlist would most likely become game-specific software that the developer's would host for their games and related content.

Even more user distribution? No. Absolutely against it (not that I’m in a position to decide anything). This would basically kill Minetest and what was achieved till now.

> We could do nothing at all.

Or the devs could provide an actual game shipped with the default installation instead of the non-game called “Minetest Game” that is good for nothing except for being used as modding base.

**Comment by ClobberXD** [comment]
Interesting idea. I don't know much about Lua runtime environments like Love2D, so I can't say much about them, but here are my thoughts about the rest of the idea and its potential implications:

I for one have always wished for a specialized client for CTF. While MT is quite good from a general overall aspect, a PvP game like CTF would have a lot more combat-oriented bells and whistles, and would discard unrelated game mechanics like creative mode, mapgen, etc.

MT is too high-level to be referred to, as a game engine. It's more like a web browser that renders web-pages (games) containing javascript (mods). I mean, MT isn't a game engine in a purist sense - it does way more than just provide a rendering engine, handle I/O and networking, etc. It goes as far as to provide an arguably restricted API, thereby limiting the creative ability of the "games" that run on it.

On the other hand, MT was created from the ground-up (with full knowledge of the aforementioned drawback) to support extensive modding/scripting (and it really has done quite well on this front). With an easy-to-use scripting language like Lua, it's just better to make use of an existing (albeit restrictive) API, rather than just use MT as a rendering engine, and define all other game mechanics in games/mods - that'd be too much to do in a scripting language.

Another purpose of scripting is that many more people can create content. Content creators, like everyone else, find Lua much easier than C++, and would effectively be able to create much more content in the form of mods, rather than int the form of a full-blown game in C++ based on a voxel game engine called Minetest.

 **The strength of Minetest is its community**. Supporting custom clients for games would mean that the Minetest community would suddenly cease to exist - hordes of people would join the communities of their respective games, and only a small bunch of people (that too mostly devs) would stick around. Speaking of devs, why would they want to put their precious time and effort to help improve/maintain a dead project? MT would stagnate, and then slowly vanish into the darkness...

TL;DR: While I like the idea of distributing custom clients for games, I think Minetest should remain as it is, and not encourage each game to make their own client, because:

- Creating games in C++ is no joke, and takes way more effort than it's generally worth.
- Again, the community.

**Comment by DS-Minetest** [comment]
Minetest is lacking dev time, not only in the engine but also for mods, subgames and other content. It would be an immense effort for every subgame to create an own builtin.
Also the different subgames don't differ that much and the main menu is not really part of the game. You can join servers and co.. It's more like a content starter.

The following might be a bit offtopic. In LÖVE, you can zip your lua game and change the file suffix to `.love`, then you can call love on this file to interpret and start the game. Something like this for minetest might be interesting, too. For example zip a world folder and rename it from `my_world.zip` to `my_world.world.minetest` and call minetest on it to put it automatically into the world folder (and maybe also delete the file and start minetest). ([Adding content manually can be difficult if you are new to minetest.](https://www.youtube.com/watch?v=nI8Q1bqT8QU&feature=youtu.be&t=61))

**Comment by paramat** [comment]
I don't think MT acts like a game, it acts like a game engine.
> This can be problematic for game developers who want their game to be respected as independent,

I can't see a problem here, games are already respected as independent.
> or who want further control over the client configuration

This can be attended to in other ways by allowing a game more control over clients.
> subgames were renamed to games in order to further the idea that they are independent projects and more than just large mods, but they still live within the Minetest ecosystem

I can't see a significant problem.
> games made with MT will always require users to install MT first,

I can't see a significant problem, installing a game engine first before installing a game is reasonable.
> The CDB and serverlist would most likely become game-specific software that the developer's would host for their games and related content.

This makes it more difficult to create games, currently almost anyone can, which is good.

Overall i don't see any significant issues, but your suggested solutions are extremely big structural changes, so this is not worth doing, especially considering that we will almost always have a critical lack of core dev time.
MT may not be structured 
> We could do nothing at all. We could keep this strange mixture of Game and Engine,

We're not going to that either, multiple games are intended and i hope that will happen soon.

**Comment by benrob0329** [comment]
>No, it doesn’t. It is a game engine.

Well yes but actually no. Minetest is a sort of "meta game", its almost a content distribution platform but still clings to its game-lile roots.

For example, all `games` use the same menu, are all treated the same by the serverlist, all require the same parent program to be installed separately, and all are essentially big modpacks which define and override default mapgens, behaviours, and settings (except default client-side settings, unless played in singleplayer). Mods are assumed to be compatible with MTG by default, and don't get to be separate from mods for other `games`.

The closest things to it is Garry's Mod, Source Engine Mods, Roblox, Minecraft, and other games which either allow large content packs or different game-modes.

>This would basically kill Minetest and what was achieved till now.

It would probably kill the MTG-centric modding community we have now, but would also allow each `game` to form a small community of its own separate from MTG or other `games`.


>But the engine gets shipped with a boring and uninspired non-game as default.
>Or the devs could provide an actual game shipped with the default installation instead of the non-game called “Minetest Game” that is good for nothing except for being used as modding base.

Personally, I'm of the opinion that no `game`    should be shipped by default at all, but rather that users should be shown the ContentDB as first launch to be able to see what kind of variety there is to choose from.

**Comment by paramat** [comment]
For this discussion, let's assume MT is shipped with multiple games, as intended, so let's not get distracted by it being MTG-only or people's opinions of that game.

**Comment by benrob0329** [comment]
>Creating games in C++ is no joke, and takes way more effort than it's generally worth.

I never said anything about changing the language or requiring people to fork, **in fact those are both things I would specifically like not to happen.** Game engines like Ogre, Love2D, and Unreal don't require users to fork the engine to customize things and have a game separate from a central platform.

>The following might be a bit offtopic. In LÖVE, you can zip your lua game and change the file suffix to .love, then you can call love on this file to interpret and start the game. Something like this for minetest might be interesting, too. 

Packing games like LÖVE allows would be nice, something separated from a central menu but still contained inside the Lua environment that _can_ be shipped with the engine on platforms like Windows.

>I don't think MT acts like a game, it acts like a game framework.

Again, yes but actually no. It acts like a game framework with most of the API, and with the flexibility of the engine. However, all `games` must be selected from inside the engine, which is not at all like a game framework. A game framework would allow you to launch each game independently from the system menu, and allow distribution outside of a central launcher.

>I can't see a significant problem, installing a game engine first before installing a game is reasonable.

Well yes, but that's majorly oversimplifying the process. Normally, the game engine is shipped as a library or backing framework that the user never sees. Its launched when the user wants to play a game built with it. People don't "play" the engine, they play games built with the engine.

Minetest isn't like this.


I'd like to note that both being a content distribution platform (what MT currently seems to be becoming) and being a formal game engine/framework have up sides and downsides. **This issue is to discuss both, and attempt to solidify which way Minetest is going and how games should be represented in the future.**

**Comment by LoneWolfHT** [comment]
> Personally, I'm of the opinion that no game should be shipped by default at all, but rather that users should be shown the ContentDB as first launch to be able to see what kind of variety there is to choose from.

:+1:

**Comment by rubenwardy** [comment]
I support shipping games as a third party product. One of my aims with improving the main menu is to sandbox and improve it such that games can have customised main menus

I introduced formspec prepends to help remove superfluous dependencies on default. I see default as a major obstacle to the minetest mission - it makes multi-game modding very difficult. I would like the time and support to split it up into different mods, but such a thing has been quite opposed

**Comment by paramat** [comment]
benrob0329, yes i accept that MT probably doesn't behave like most game engines or frameworks. However this isn't a significant problem, it may not be to the taste of some, but that's how it is and it's too much disruption and work to make a big change. We can however, gradually go in a certain direction as MT evolves.

The issue of default isn't that splitting it up is possible but has been opposed by others, it simply just isn't possible without either breakage of mods or creating an even more horrible mess than before.

Besides, we shouldn't be focussing on MTG and default so much, the more we focus on it the  longer we procastinate and put off the inevitable and necessary step of moving away from it and creating new mod base games with a good structure.
Continuing to try to perfect MTG is laziness and avoidance of creating new games, and just perpetuates the pain.

While i'm ok with occasional refinement and minor improvements, i think we should leave MTG mostly as it is and only maintain and bugfix it. Most of our 'game development time' should be spent on new games.
MTG can never be the super-impressive game to impress new players, and it can never be a good mod base, and it certainly cannot be both. Because: It is so limited by how much depends on it being the way it is. It does many things in stupid ways, that can't be changed, and it has a bad structure, that can't be changed.

**Comment by benrob0329** [comment]
paramat: This is about what I anticipated, but I figured I'd bring it up since games (specifically Blockcolor in recent memory) *are* trying to act like fully separate games.

**But** I think an important point was brought up, launching games outside the in-game menu. If Minetest is to become a content distribution platform (the ContentDB and its frontend at the very least) then I think it would be in the best interest to allow players to launch games as if they are separate, with their own menus and everything (even if under the hood, its just Minetest loading up a specific archive or folder). This way, we lose the "subgame" feeling, but still keep a closeknit community.

Alternatively, we could do the same thing from the initial menu shown when launching Minetest.

I also feel that both the CDB and the Forums need to allow mods to be specific to games, perhaps specifying which game(s) a mod is compatible with and allowing users to browse mods on a per-gane basis.

**Comment by LoneWolfHT** [comment]
> I also feel that both the CDB and the Forums need to allow mods to be specific to games, perhaps specifying which game(s) a mod is compatible with and allowing users to browse mods on a per-gane basis.

`game_depends`
`optional_game_depends`

**Comment by rubenwardy** [comment]
The problem with depending on a specific game is that mods never depend on a game, they depend on the mods in a game. Games are a collection of mods and have no content by themselves

CDB already supports depending on games, btw, but it's not advised.

**Comment by fraang** [comment]
Intrestingly OpenRA has killed off their in-game mod switcher not so long ago and uses now seperate menu entries. Each of the "mods" can use their own version of the engine including forks. This was done to make the life of devs easier, so they don't have to keep their mods in sync with engine developments (right away when they make breaking changes in the engine) and can do it in their own way/speed. I am not sure if that is good thing or not. It cerainly makes it harder to look trough the "version jungle". :-(

**Comment by hecktest** [comment]
Don't change. The minetest architecture reminds me of MUDs in a good way. Not having to install anything on top of MT is a Good Thing.
A shared player base provides a positive feedback loop. I can launch a server, and have guaranteed players to populate it. Then, someone might download MT to play my server/subgame, and try others too since the effort involved in that is minimal. If I end up fixing a blocker bug in the engine, I'm more or less forced to make a PR for it, otherwise it won't count.
![sub](https://user-images.githubusercontent.com/42101236/69083900-5433f800-0a43-11ea-80a8-282142b85a1a.jpg)
Now force me to host a separate client, and the opposite will be true; you'll bleed users and gain nothing, as I'd have no incentive to maintain engine compatibility. Not to mention all the wasted, duplicated work.

Of course, there are things that can be improved. Give us separate menus and configs, sure. Carve out more MTG-specific things to lua, so that I don't have to sidestep them. But keep the "terminal to play games over the internet" model, it's the whole point of this thing.
